src/Controllers/MessagesController.php

```diff
@@ -0,0 +1,285 @@
+        }
+
+        $new_message = new Message();
+        $new_message->sender()->associate($current_user);
+        $new_message->receiver()->associate($other_user);
+        $new_message->text = $data['text'];
+        $new_message->read = false;
+        $new_message->save();
+
+        $this->log->info(
+            'User {from} has written a message to user {to}',
+            [
+                'from' => $current_user->id,
+                'to' => $other_user->id
+            ]
+        );
+        return $this->redirect->to('/messages/' . $other_user->id);
+    }
+
+    /**
+     * Deletes a message from a given id, as long as this message was send by the current user. The given user_id
+     * The given user_id is used to redirect back to the conversation with that user.
+     */
+    public function delete(Request $request): Response
+    {
+        $current_user = $this->auth->user();
+        $other_user_id = $request->getAttribute('user_id');
+        $msg_id = $request->getAttribute('msg_id');
+        $msg = $this->message->findOrFail($msg_id);
+
+        if ($msg->user_id == $current_user->id) {
+            $msg->delete();
+
+            $this->log->info(
+                'User {from} deleted message {msg} in a conversation with user {to}',
+                [
+                    'from' => $current_user->id,
+                    'to' => $other_user_id,
+                    'msg' => $msg_id
+                ]
+            );
+        } else {
+            $this->log->warning(
+                'User {from} tried to delete message {msg} which was not written by them, ' .
+                'in a conversation with user {to}',
+                [
+                    'from' => $current_user->id,
+                    'to' => $other_user_id,
+                    'msg' => $msg_id
+                ]
+            );
+
+            throw new HttpForbidden('You can not delete a message you haven\'t send');
+        }
+
+        return $this->redirect->to('/messages/' . $other_user_id);
+    }
+
+    public function numberOfUnreadMessages(): int
+    {
+        return $this->auth->user()
+            ->messagesReceived()
+            ->where('read', false)
+            ->count();
+    }
+
+    protected function numberOfUnreadMessagesPerConversation($current_user): Collection
+    {
+        return $current_user->messagesReceived()
+            ->select('user_id', $this->raw('count(*) as amount'))
+            ->where('read', false)
+            ->groupBy('user_id')
+            ->get(['user_id', 'amount'])
+            ->mapWithKeys(function ($unread) {
+                return [ $unread->user_id => $unread->amount ];
+            });
+    }
+
+    protected function latestMessagePerConversation($current_user): Collection
+    {
+        $latest_message_ids = $this->message
+            ->select($this->raw('max(id) as last_id'))
+            ->where('user_id', "=", $current_user->id)
```

Comment:
Eventually using just `'` instead of a mix with `"` if no substitution is required is a great idea